src/server/pkg/pool/pool.go

```diff
@@ -70,65 +68,72 @@ func (p *Pool) watchEndpoints() {
 }
 
 func (p *Pool) updateAddresses(endpoints *api.Endpoints) {
-	var addresses []string
+	addresses := make(map[string]*connCount)
+	p.connsLock.Lock()
+	defer p.connsLock.Unlock()
 	for _, subset := range endpoints.Subsets {
 		// According the k8s docs, the full set of endpoints is the cross
 		// product of (addresses x ports).
 		for _, address := range subset.Addresses {
 			for _, port := range subset.Ports {
-				addresses = append(addresses, fmt.Sprintf("%s:%d", address.IP, port.Port))
+				addr := fmt.Sprintf("%s:%d", address.IP, port.Port)
+				if cc := p.conns[addr]; cc != nil {
+					addresses[addr] = cc
+				} else {
+					addresses[addr] = nil
+				}
 			}
 		}
 	}
-	p.addressesLock.Lock()
-	defer p.addressesLock.Unlock()
-	p.addresses = addresses
-	p.addressesCond.Broadcast()
+	p.conns = addresses
 }
 
-// Get returns a new connection, unlike sync.Pool if it has a cached connection
-// it will always return it. Otherwise it will create a new one. Get errors
-// only when it needs to Dial a new connection and that process fails.
-func (p *Pool) Get(ctx context.Context) (*grpc.ClientConn, error) {
-	select {
-	case conn := <-p.conns:
-		return conn, nil
-	default:
-		p.addressesLock.Lock()
-		defer p.addressesLock.Unlock()
-		for len(p.addresses) == 0 {
-			p.addressesCond.Wait()
+// Do allows you to do something with a grpc.ClientConn.
+// Errors returned from f will be returned by Do.
+func (p *Pool) Do(ctx context.Context, f func(cc *grpc.ClientConn) error) error {
+	var conn *connCount
+	if err := func() error {
+		p.connsLock.Lock()
+		defer p.connsLock.Unlock()
+		for addr, mapConn := range p.conns {
+			if mapConn == nil {
+				cc, err := grpc.DialContext(ctx, addr, p.opts...)
+				if err != nil {
+					return err
+				}
+				conn = &connCount{cc: cc}
+				p.conns[addr] = conn
+				// We break because this conn has a count of 0 which we know
+				// we're not beating
+				break
+			} else {
+				if conn == nil || mapConn.count < conn.count {
```

Comment:
I think for atomic variables, you are supposed to use an `atomic.Load*` function to get the value of it.